File: implement/b2630/solution.py
import sys

input = sys.stdin.readline
n = int(input())
graph = [list(map(int, input().split())) for _ in range(n)]

white = 0
blue = 0


def quad_tree(start_x, start_y, graph_size):
    global white
    global blue

    cur = graph[start_x][start_y]
    for i in range(start_x, start_x + graph_size):
        for j in range(start_y, start_y + graph_size):
            if cur != graph[i][j]:
                next_size = graph_size // 2
                quad_tree(start_x, start_y, next_size)  # 왼쪽 위
                quad_tree(start_x, start_y + next_size, next_size)  # 오른쪽 위
                quad_tree(start_x + next_size, start_y, next_size)  # 왼쪽 아래
                quad_tree(start_x + next_size, start_y + next_size, next_size)  # 오른쪽 아래

                return

            cur = graph[i][j]

    if cur == 0:
        white += 1
    else:
        blue += 1
    return


quad_tree(0, 0, n)

print(white)
print(blue)

File: implement/b2630/Solution.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Arrays;

public class Solution {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int[][] graph;

    static int white = 0;
    static int blue = 0;

    public static void main(String[] args) throws IOException {
        int n = Integer.parseInt(br.readLine());
        graph = new int[n][n];

        for (int i = 0; i < n; i++) {
            graph[i] = Arrays.stream(br.readLine().split(" "))
                    .mapToInt(Integer::parseInt)
                    .toArray();
        }

        quadTree(0, 0, n);
        print();
    }

    static void quadTree(int x, int y, int graphSize) {
        int cur = graph[x][y];

        for (int i = x; i < x + graphSize; i++) {
            for (int j = y; j < y + graphSize; j++) {
                if (cur != graph[i][j]) {
                    int nextSize = graphSize / 2;

                    quadTree(x, y, nextSize);
                    quadTree(x, y + nextSize, nextSize);
                    quadTree(x + nextSize, y, nextSize);
                    quadTree(x + nextSize, y + nextSize, nextSize);
                    return;
                }

                cur = graph[i][j];

            }
        }

        if (cur == 0) {
            white++;
        } else {
            blue++;
        }

        return;
    }

    static void print() throws IOException {
        bw.write(white + "\n");
        bw.write(blue + "\n");
        bw.flush();

        br.close();
        bw.close();
    }
}
